### PR TITLE
Add `CyclotomicField` to represent cyclotomic operations

### PR DESCRIPTION
**Pull request by Pratyush**
## Description

Another part of https://github.com/arkworks-rs/algebra/pull/294

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [x] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

File: ff/src/biginteger/arithmetic.rs [hunk 1]
```diff
@@ -67,7 +67,7 @@ pub(crate) fn mac_discard(a: u64, b: u64, c: u64, carry: &mut u64) {
 }
 
 /// Compute the NAF (non-adjacent form) of num
-pub fn find_naf(num: &[u64]) -> Vec<i64> {
+pub fn find_naf(num: &[u64]) -> Vec<i8> {
```
Comment (ValarDragon) [inline ff/src/biginteger/arithmetic.rs:70]: naf change LGTM, but a bit confused why were bundling it into this PR

Comment (Pratyush) [reply ff/src/biginteger/arithmetic.rs:70]: Got pulled in from the previous PR (#294)

Comment (Pratyush) [reply ff/src/biginteger/arithmetic.rs:70]: (I can push it into its own PR if that's better)

File: ff/src/fields/mod.rs [hunk 2]
```diff
@@ -268,6 +268,102 @@ pub trait Field:
     }
 }
 
+/// Fields that have a cyclotomic multiplicative subgroup, and which can
+/// leverage efficient inversion and squaring algorithms for elements in this subgroup.
+/// If a field has multiplicative order p^d - 1, the cyclotomic subgroups refer to subgroups of order φ_n(p),
+/// for any n < d, where φ_n is the [nth cyclotomic polynomial](https://en.wikipedia.org/wiki/Cyclotomic_polynomial)
+pub trait CyclotomicField: Field {
```
Comment (ValarDragon) [inline ff/src/fields/mod.rs:275]: Thoughts on naming the trait `CyclotomicSubgroup`, to better highlight this is improved arithmetic for points that are in a subgroup, not the entire field?

Comment (Pratyush) [reply ff/src/fields/mod.rs:275]: That's a good idea. I'd like to find to do that while avoiding confusion with the rest of the changes coming in from #294, which re-introduce an _additive_ group abstraction (while this is a _multiplicative_ group abstraction)

File: ff/src/fields/models/fp12_2over3over2.rs [hunk 3]
```diff
@@ -230,6 +206,95 @@ pub fn characteristic_square_mod_6_is_one(characteristic: &[u64]) -> bool {
...
+    const INVERSE_IS_FAST: bool = true;
+
+    fn cyclotomic_inverse_in_place(&mut self) -> Option<&mut Self> {
+        self.is_zero().not().then(|| {
+            self.conjugate();
+            self
+        })
+    }
+
+    fn cyclotomic_square_in_place(&mut self) -> &mut Self {
```
Comment (mmagician) [inline ff/src/fields/models/fp12_2over3over2.rs:219]: The method is now duplicated in `impl Fp12` and `impl CyclotomicField for Fp12` - it should rather be unified no?

**Comment by mmagician** [comment]
Adding (doc?) examples for using `square` from Field & `cyclotomic_square` would further help to avoid confusion.

File: ff/src/fields/models/fp12_2over3over2.rs (hunk 3 above)
Comment (Pratyush) [reply ff/src/fields/models/fp12_2over3over2.rs:219]: Fixed

Comment (Pratyush) [reply ff/src/fields/models/fp12_2over3over2.rs:219]: Fixed!

File: ff/src/fields/mod.rs (hunk 2 above)
Comment (Pratyush) [reply ff/src/fields/mod.rs]: Renamed to `CyclotomicMultSubgroup`, and added a note specifying that this is different from `ark_ec::Group`.

File: ff/src/biginteger/arithmetic.rs (hunk 1 above)
Comment (Pratyush) [reply ff/src/biginteger/arithmetic.rs:70]: For the time being I've left it in, since it's a small change.

**Review by mmagician — APPROVED** [review]

**Pull request merged**